src/queue.c

```diff
@@ -5594,71 +5638,90 @@ _dispatch_runloop_root_queue_wakeup_4CF(dispatch_queue_t dq)
 	_dispatch_runloop_queue_wakeup(dq, 0, false);
 }
 
-mach_port_t
+dispatch_runloop_handle_t
 _dispatch_runloop_root_queue_get_port_4CF(dispatch_queue_t dq)
 {
 	if (slowpath(dq->do_vtable != DISPATCH_VTABLE(queue_runloop))) {
 		DISPATCH_CLIENT_CRASH(dq->do_vtable, "Not a runloop queue");
 	}
-	return (mach_port_t)dq->do_ctxt;
+	return _dispatch_runloop_queue_get_handle(dq);
 }
 
 static void
-_dispatch_runloop_queue_port_init(void *ctxt)
+_dispatch_runloop_queue_handle_init(void *ctxt)
 {
 	dispatch_queue_t dq = (dispatch_queue_t)ctxt;
-	mach_port_t mp;
-	kern_return_t kr;
+	dispatch_runloop_handle_t handle;
 
 	_dispatch_fork_becomes_unsafe();
-	kr = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE, &mp);
+
+#if TARGET_OS_MAC
+	kern_return_t kr;
+	kr = mach_port_allocate(mach_task_self(), MACH_PORT_RIGHT_RECEIVE, &handle);
 	DISPATCH_VERIFY_MIG(kr);
 	(void)dispatch_assume_zero(kr);
-	kr = mach_port_insert_right(mach_task_self(), mp, mp,
+	kr = mach_port_insert_right(mach_task_self(), handle, handle,
 			MACH_MSG_TYPE_MAKE_SEND);
 	DISPATCH_VERIFY_MIG(kr);
 	(void)dispatch_assume_zero(kr);
 	if (dq != &_dispatch_main_q) {
 		struct mach_port_limits limits = {
 			.mpl_qlimit = 1,
 		};
-		kr = mach_port_set_attributes(mach_task_self(), mp,
+		kr = mach_port_set_attributes(mach_task_self(), handle,
 				MACH_PORT_LIMITS_INFO, (mach_port_info_t)&limits,
 				sizeof(limits));
 		DISPATCH_VERIFY_MIG(kr);
 		(void)dispatch_assume_zero(kr);
 	}
-	dq->do_ctxt = (void*)(uintptr_t)mp;
+#elif defined(__linux__)
+	handle = eventfd(0, EFD_CLOEXEC | EFD_NONBLOCK);
+#else
+#error "runloop support not implemented on this platform"
+#endif	
+	(void)dispatch_assume(_dispatch_runloop_handle_is_valid(handle));
+	_dispatch_runloop_queue_set_handle(dq, handle);
```

Comment:
do you really want to set the handle if it is invalid ? if seems like this wants to be inside an if for the assume above ?

alternatively add a hard failure (CRASH macro) here, similarly to what we have in `_dispatch_kq_init()` when we cannot create the kq file descriptor (e.g. due the process or system being out of fd's)